app/controllers/api/base_controller.rb

```diff
@@ -231,6 +232,20 @@ def not_found_if_nested_id_exists
       end
     end
 
+    def missing_permissions
+      Foreman::AccessControl.permissions_for_controller_action(path_to_authenticate)
+    end
+
+    def set_error_details(error, options)
+      case error
+      when 'access_denied'
+        if options.is_a?(Hash) && options.fetch(:locals, {}).fetch(:details, nil).blank?
```

Comment:
When would `options` not be a hash? Both `render_error` methods rely on it to be a hash to set the template/layout.

I also wonder why not appending the error to `[:locals][:details]` directly even if it contains something?